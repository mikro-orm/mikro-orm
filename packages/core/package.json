{
  "name": "@mikro-orm/core",
  "version": "6.1.12",
  "description": "TypeScript ORM for Node.js based on Data Mapper, Unit of Work and Identity Map patterns. Supports MongoDB, MySQL, PostgreSQL and SQLite databases as well as usage with vanilla JavaScript.",
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "typings": "dist/index.d.ts",
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "import": {
        "types": "./dist/index.d.ts",
        "default": "./dist/index.mjs"
      },
      "require": "./dist/index.js"
    }
  },
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com/mikro-orm/mikro-orm.git"
  },
  "funding": "https://github.com/sponsors/b4nan",
  "keywords": [
    "orm",
    "mongo",
    "mongodb",
    "mysql",
    "mariadb",
    "postgresql",
    "sqlite",
    "sqlite3",
    "ts",
    "typescript",
    "js",
    "javascript",
    "entity",
    "ddd",
    "mikro-orm",
    "unit-of-work",
    "data-mapper",
    "identity-map"
  ],
  "author": "Martin AdÃ¡mek",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/mikro-orm/mikro-orm/issues"
  },
  "homepage": "https://mikro-orm.io",
  "engines": {
    "node": ">= 18.12.0"
  },
  "scripts": {
    "build": "yarn clean && yarn compile && yarn copy && yarn run -T gen-esm-wrapper dist/index.js dist/index.mjs",
    "clean": "yarn run -T rimraf ./dist",
    "compile": "yarn run -T tsc -p tsconfig.build.json",
    "copy": "node ../../scripts/copy.mjs"
  },
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "dataloader": "2.2.2",
    "dotenv": "16.4.5",
    "esprima": "4.0.1",
    "fs-extra": "11.2.0",
    "globby": "11.1.0",
    "mikro-orm": "6.1.12",
    "reflect-metadata": "0.2.2"
  }
}
