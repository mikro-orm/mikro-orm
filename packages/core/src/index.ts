/**
 * @packageDocumentation
 * @module core
 */
/* istanbul ignore file */
export {
  Constructor, Dictionary, PrimaryKeyType, PrimaryKeyProp, Primary, IPrimaryKey, FilterQuery, IWrappedEntity, EntityName, EntityData, Highlighter,
  AnyEntity, EntityProperty, EntityMetadata, QBFilterQuery, PopulateOptions, Populate, Loaded, New, LoadedReference, LoadedCollection,
  GetRepository, EntityRepositoryType, MigrationObject, DeepPartial, PrimaryProperty, Cast, IsUnknown, EntityDictionary, EntityDTO, PlainObject,
} from './typings';
export * from './enums';
export * from './errors';
export * from './exceptions';
export * from './MikroORM';
export * from './entity';
export * from './events';
export * from './EntityManager';
export * from './unit-of-work';
export * from './utils';
export * from './hydration';
export * from './drivers';
export * from './connections';
export * from './platforms';
export * from './types';
export * from './naming-strategy';
export * from './metadata';
export * from './cache';
export * from './decorators';
