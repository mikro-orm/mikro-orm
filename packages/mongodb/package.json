{
  "name": "@mikro-orm/mongodb",
  "version": "5.2.4",
  "description": "TypeScript ORM for Node.js based on Data Mapper, Unit of Work and Identity Map patterns. Supports MongoDB, MySQL, PostgreSQL and SQLite databases as well as usage with vanilla JavaScript.",
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "typings": "dist/index.d.ts",
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "import": {
        "types": "./dist/index.d.ts",
        "default": "./dist/index.mjs"
      },
      "require": "./dist/index.js"
    }
  },
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com/mikro-orm/mikro-orm.git"
  },
  "keywords": [
    "orm",
    "mongo",
    "mongodb",
    "mysql",
    "mariadb",
    "postgresql",
    "sqlite",
    "sqlite3",
    "ts",
    "typescript",
    "js",
    "javascript",
    "entity",
    "ddd",
    "mikro-orm",
    "unit-of-work",
    "data-mapper",
    "identity-map"
  ],
  "author": "Martin AdÃ¡mek",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/mikro-orm/mikro-orm/issues"
  },
  "homepage": "https://mikro-orm.io",
  "engines": {
    "node": ">= 14.0.0"
  },
  "scripts": {
    "build": "yarn clean && yarn compile && yarn copy",
    "postbuild": "yarn gen-esm-wrapper dist/index.js dist/index.mjs",
    "clean": "rimraf ./dist",
    "compile": "tsc -p tsconfig.build.json",
    "copy": "ts-node -T ../../scripts/copy.ts"
  },
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "bson": "^4.6.5",
    "mongodb": "4.8.1"
  },
  "devDependencies": {
    "@mikro-orm/core": "^5.2.4"
  },
  "peerDependencies": {
    "@mikro-orm/core": "^5.0.0",
    "@mikro-orm/entity-generator": "^5.0.0",
    "@mikro-orm/migrations": "^5.0.0",
    "@mikro-orm/seeder": "^5.0.0"
  },
  "peerDependenciesMeta": {
    "@mikro-orm/entity-generator": {
      "optional": true
    },
    "@mikro-orm/migrations": {
      "optional": true
    },
    "@mikro-orm/seeder": {
      "optional": true
    }
  }
}
