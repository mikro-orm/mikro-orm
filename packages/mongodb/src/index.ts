/* istanbul ignore file */
export * from './MongoConnection';
export * from './MongoDriver';
export * from './MongoPlatform';
export * from './MongoEntityManager';
export * from './MongoEntityRepository';
export * from './MongoSchemaGenerator';
export { MongoEntityManager as EntityManager } from './MongoEntityManager';
export { MongoEntityRepository as EntityRepository } from './MongoEntityRepository';
export {
  MongoMikroORM as MikroORM,
  MongoOptions as Options,
  defineMongoConfig as defineConfig,
} from './MongoMikroORM';
export { ObjectId } from 'bson';
export * from '@mikro-orm/core';
