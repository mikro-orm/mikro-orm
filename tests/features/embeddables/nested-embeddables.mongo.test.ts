import { Embeddable, Embedded, Entity, PrimaryKey, Property } from '@mikro-orm/core';
import { MikroORM, ObjectId, MongoConnection } from '@mikro-orm/mongodb';
import { mockLogger } from '../../helpers';

@Embeddable()
class IdentityMeta {

  @Property()
  foo?: string;

  @Property({ index: true })
  bar?: string;

  constructor(foo?: string, bar?: string) {
    this.foo = foo;
    this.bar = bar;
  }

}

@Embeddable()
class Identity {

  @Property({ unique: true })
  email: string;

  @Embedded(() => IdentityMeta, { nullable: true })
  meta?: IdentityMeta;

  constructor(email: string, meta?: IdentityMeta) {
    this.email = email;
    this.meta = meta;
  }

}

@Embeddable()
class Profile {

  @Property({ unique: true })
  username: string;

  @Embedded(() => Identity)
  identity: Identity;

  constructor(username: string, identity: Identity) {
    this.username = username;
    this.identity = identity;
  }

}

@Entity()
class User {

  @PrimaryKey()
  _id!: ObjectId;

  @Property()
  name!: string;

  @Embedded(() => Profile)
  profile1!: Profile;

  @Embedded(() => Profile, { object: true })
  profile2!: Profile;

}

describe('embedded entities in mongo', () => {

  let orm: MikroORM;

  beforeAll(async () => {
    orm = await MikroORM.init({
      entities: [User],
      clientUrl: 'mongodb://localhost:27017/mikro-orm-test-nested-embeddables',
    });
  });

  beforeEach(async () => {
    await orm.schema.clearDatabase();
  });

  afterAll(async () => {
    await orm.schema.dropSchema({ dropMigrationsTable: true });
    await orm.close(true);
  });

  test('create collections', async () => {
    const createCollection = jest.spyOn(MongoConnection.prototype, 'createCollection');
    createCollection.mockResolvedValue({} as any);
    await orm.schema.createSchema();
    expect(createCollection.mock.calls.map(c => c[0])).toEqual(['user', 'mikro_orm_migrations']);
    createCollection.mockRestore();
  });

  test('create nested indexes', async () => {
    await orm.schema.ensureIndexes();
    const userInfo = await orm.em.getCollection(User).indexInformation({ full: true, session: undefined as any });
    expect(userInfo.reduce((o: any, i: any) => { o[i.name] = i; return o; }, {} as any)).toMatchObject({
      '_id_': {
        key: { _id: 1 },
      },
      'profile1_identity_email_1': {
        key: { profile1_identity_email: 1 },
        sparse: false,
        unique: true,
      },
      'profile1_identity_meta_bar_1': {
        key: { profile1_identity_meta_bar: 1 },
        sparse: true,
      },
      'profile1_username_1': {
        key: { profile1_username: 1 },
        sparse: false,
        unique: true,
      },
      'profile2.identity.email_1': {
        key: { 'profile2.identity.email': 1 },
        sparse: false,
        unique: true,
      },
      'profile2.identity.meta.bar_1': {
        key: { 'profile2.identity.meta.bar': 1 },
        sparse: true,
      },
      'profile2.username_1': {
        key: { 'profile2.username': 1 },
        sparse: false,
        unique: true,
      },
    });
  });

  test('diffing', async () => {
    expect(orm.em.getComparator().getSnapshotGenerator('User').toString()).toMatchSnapshot();
  });

  test('unique constraints', async () => {
    await orm.schema.ensureIndexes();

    const user1 = new User();
    user1.name = 'Uwe';
    user1.profile1 = new Profile('u1', new Identity('e1', new IdentityMeta('f1', 'b1')));
    user1.profile2 = new Profile('u2', new Identity('e2', new IdentityMeta('f2', 'b2')));

    const user2 = new User();
    user2.name = 'Uschi';
    user2.profile1 = new Profile('u1', new Identity('e3'));
    user2.profile2 = new Profile('u4', new Identity('e4', new IdentityMeta('f4')));

    await expect(orm.em.persistAndFlush([user1, user2])).rejects.toThrowError(/E11000 duplicate key error collection: mikro-orm-test-nested-embeddables\.user index: profile1_username_1 dup key: \{ profile1_username: "u1" }/);
  });

  test('persist and load', async () => {
    const user1 = new User();
    user1.name = 'Uwe';
    user1.profile1 = new Profile('u1', new Identity('e1', new IdentityMeta('f1', 'b1')));
    user1.profile2 = new Profile('u2', new Identity('e2', new IdentityMeta('f2', 'b2')));

    const user2 = new User();
    user2.name = 'Uschi';
    user2.profile1 = new Profile('u3', new Identity('e3'));
    user2.profile2 = new Profile('u4', new Identity('e4', new IdentityMeta('f4')));

    const mock = mockLogger(orm);
    await orm.em.persistAndFlush([user1, user2]);
    orm.em.clear();
    expect(mock.mock.calls[0][0]).toMatch(`db.getCollection('user').insertMany([ { name: 'Uwe', profile1_username: 'u1', profile1_identity_email: 'e1', profile1_identity_meta_foo: 'f1', profile1_identity_meta_bar: 'b1', profile2: { username: 'u2', identity: { email: 'e2', meta: { foo: 'f2', bar: 'b2' } } } }, { name: 'Uschi', profile1_username: 'u3', profile1_identity_email: 'e3', profile2: { username: 'u4', identity: { email: 'e4', meta: { foo: 'f4' } } } } ], {});`);

    const u1 = await orm.em.findOneOrFail(User, user1._id);
    const u2 = await orm.em.findOneOrFail(User, user2._id);
    expect(mock.mock.calls[1][0]).toMatch(/db\.getCollection\('user'\)\.find\({ _id: .* }, {}\)\.limit\(1\).toArray\(\);/);
    expect(u1.profile1).toBeInstanceOf(Profile);
    expect(u1.profile1).toEqual({
      username: 'u1',
      identity: {
        email: 'e1',
        meta: {
          bar: 'b1',
          foo: 'f1',
        },
      },
    });
    expect(u1.profile2).toBeInstanceOf(Profile);
    expect(u1.profile2).toEqual({
      username: 'u2',
      identity: {
        email: 'e2',
        meta: {
          bar: 'b2',
          foo: 'f2',
        },
      },
    });
    expect(u2.profile1).toBeInstanceOf(Profile);
    expect(u2.profile1).toEqual({
      username: 'u3',
      identity: {
        email: 'e3',
      },
    });
    expect(u2.profile2).toBeInstanceOf(Profile);
    expect(u2.profile2).toEqual({
      username: 'u4',
      identity: {
        email: 'e4',
        meta: {
          foo: 'f4',
        },
      },
    });

    expect(mock.mock.calls.length).toBe(3);
    await orm.em.flush();
    expect(mock.mock.calls.length).toBe(3);

    u1.profile1!.identity.email = 'e123';
    u1.profile1!.identity.meta!.foo = 'foooooooo';
    u1.profile2!.identity.meta!.bar = 'bababar';
    await orm.em.flush();
    expect(mock.mock.calls[3][0]).toMatch(/db\.getCollection\('user'\)\.updateMany\({ _id: .* }, { '\$set': { profile1_identity_email: 'e123', profile1_identity_meta_foo: 'foooooooo', profile2: { username: 'u2', identity: { email: 'e2', meta: { foo: 'f2', bar: 'bababar' } } } } }, {}\);/);
    orm.em.clear();
    mock.mock.calls.length = 0;

    const u3 = await orm.em.findOneOrFail(User, {
      profile1: { identity: { email: 'e123', meta: { foo: 'foooooooo' } } },
      profile2: { identity: { email: 'e2', meta: { foo: 'f2', bar: 'bababar' } } },
    });
    expect(mock.mock.calls[0][0]).toMatch(/db\.getCollection\('user'\)\.find\({ profile1_identity_email: 'e123', profile1_identity_meta_foo: 'foooooooo', 'profile2\.identity\.email': 'e2', 'profile2\.identity\.meta\.foo': 'f2', 'profile2\.identity\.meta\.bar': 'bababar' }, {}\)\.limit\(1\).toArray\(\);/);
    expect(u3._id).toEqual(u1._id);
    orm.em.clear();
    mock.mock.calls.length = 0;

    const u4 = await orm.em.findOneOrFail(User, {
      profile1: { identity: { email: 'e123', meta: { foo: /fo+/ } } },
      profile2: { identity: { email: 'e2', meta: { foo: 'f2', bar: /(ba)+r/ } } },
    });
    expect(u4._id).toEqual(u1._id);
    expect(mock.mock.calls[0][0]).toMatch(/db\.getCollection\('user'\)\.find\({ profile1_identity_email: 'e123', profile1_identity_meta_foo: \/fo\+\/, 'profile2\.identity\.email': 'e2', 'profile2\.identity\.meta\.foo': 'f2', 'profile2\.identity\.meta\.bar': \/\(ba\)\+r\/ }, {}\)\.limit\(1\).toArray\(\);/);
    orm.em.clear();
    mock.mock.calls.length = 0;

    const u5 = await orm.em.findOneOrFail(User, { $or: [{ profile1: { identity: { meta: { foo: 'foooooooo' } } } }, { profile2: { identity: { meta: { bar: 'bababar' } } } }] });
    expect(mock.mock.calls[0][0]).toMatch(/db\.getCollection\('user'\)\.find\({ '\$or': \[ { profile1_identity_meta_foo: 'foooooooo' }, { 'profile2\.identity\.meta\.bar': 'bababar' } ] }, {}\)\.limit\(1\).toArray\(\);/);
    expect(u5._id).toEqual(u1._id);

    const err1 = `Invalid query for entity 'User', property 'city' does not exist in embeddable 'Identity'`;
    await expect(orm.em.findOneOrFail(User, { profile1: { identity: { city: 'London 1' } as any } })).rejects.toThrowError(err1);

    const err2 = `Invalid query for entity 'User', property 'city' does not exist in embeddable 'Identity'`;
    await expect(orm.em.findOneOrFail(User, { profile2: { identity: { city: 'London 1' } as any } })).rejects.toThrowError(err2);
  });

  test('#assign() works with nested embeddables', async () => {
    const jon = new User();

    orm.em.assign(jon, {
      profile1: { username: 'u1', identity: { email: 'e1', meta: { bar: 'b1', foo: 'f1' } } },
      profile2: { username: 'u2', identity: { email: 'e2', meta: { bar: 'b2', foo: 'f2' } } },
    });
    expect(jon.profile1).toMatchObject({ username: 'u1', identity: { email: 'e1', meta: { bar: 'b1', foo: 'f1' } } });
    expect(jon.profile1).toBeInstanceOf(Profile);
    expect(jon.profile1.identity).toBeInstanceOf(Identity);
    expect(jon.profile1.identity.meta).toBeInstanceOf(IdentityMeta);
    expect(jon.profile2).toMatchObject({ username: 'u2', identity: { email: 'e2', meta: { bar: 'b2', foo: 'f2' } } });
    expect(jon.profile2).toBeInstanceOf(Profile);
    expect(jon.profile2.identity).toBeInstanceOf(Identity);
    expect(jon.profile2.identity.meta).toBeInstanceOf(IdentityMeta);

    orm.em.assign(jon, { profile1: { identity: { email: 'e3' } } });
    expect(jon.profile1.username).toBe('u1');
    expect(jon.profile1.identity.email).toBe('e3');
    expect(jon.profile1.identity.meta).not.toBeUndefined();
    delete jon.profile1.identity.meta;

    orm.em.assign(jon, { profile1: { identity: { meta: { foo: 'f' } } } });
    expect(jon.profile1.identity.meta!.foo).toBe('f');
    expect(jon.profile1.identity.meta).toBeInstanceOf(IdentityMeta);

    orm.em.assign(jon, { profile1: { identity: { email: 'e4' } } }, { mergeObjectProperties: false });
    expect(jon.profile1.username).toBeUndefined();
    expect(jon.profile1.identity.email).toBe('e4');
    expect(jon.profile1.identity.meta).toBeUndefined();
  });

});
