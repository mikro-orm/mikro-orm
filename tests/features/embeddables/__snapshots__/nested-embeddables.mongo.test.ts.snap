// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`embedded entities in mongo diffing 1`] = `
"function(entity) {
  const ret = {};
  if (typeof entity._id !== 'undefined') {
    ret._id = clone(entity._id);
  }

  if (typeof entity.name !== 'undefined') {
    ret.name = entity.name;
  }

  if (entity.profile1 != null) {
    if (typeof entity.profile1.username !== 'undefined') ret.profile1_username = clone(entity.profile1.username);

    if (entity.profile1.identity != null) {
      if (typeof entity.profile1.identity.email !== 'undefined') ret.profile1_identity_email = clone(entity.profile1.identity.email);

      if (entity.profile1.identity.meta != null) {
        if (typeof entity.profile1.identity.meta.foo !== 'undefined') ret.profile1_identity_meta_foo = clone(entity.profile1.identity.meta.foo);
        if (typeof entity.profile1.identity.meta.bar !== 'undefined') ret.profile1_identity_meta_bar = clone(entity.profile1.identity.meta.bar);
      }

    }

  }

  if (entity.profile2 === null) ret.profile2 = null;
  if (entity.profile2 != null) {
    ret.profile2 = {};
    if (typeof entity.profile2.username !== 'undefined') ret.profile2.username = clone(entity.profile2.username);

    if (entity.profile2.identity === null) ret.profile2.identity = null;
    if (entity.profile2.identity != null) {
      ret.profile2.identity = {};
      if (typeof entity.profile2.identity.email !== 'undefined') ret.profile2.identity.email = clone(entity.profile2.identity.email);

      if (entity.profile2.identity.meta === null) ret.profile2.identity.meta = null;
      if (entity.profile2.identity.meta != null) {
        ret.profile2.identity.meta = {};
        if (typeof entity.profile2.identity.meta.foo !== 'undefined') ret.profile2.identity.meta.foo = clone(entity.profile2.identity.meta.foo);
        if (typeof entity.profile2.identity.meta.bar !== 'undefined') ret.profile2.identity.meta.bar = clone(entity.profile2.identity.meta.bar);
      }

    }

    ret.profile2 = cloneEmbeddable(ret.profile2);
  }

  return ret;
}"
`;
